### PR TITLE
using rfkill to unblock and block bluetooth

### PR DESCRIPTION
**Pull request by rafaelcorsi**
fix issue #61.

I decide to not only unblock de bluetooth when on, but to soft block the bluetooth when turn off.

### CODE REVIEW

**Comment by rafaelcorsi** [comment]
the same problem occurs on the switchboard-plug-bluetooth: 

https://github.com/elementary/switchboard-plug-bluetooth/blob/4d2d730995ac41df7203fc84fbdcbc1b22b93689/src/Services/Manager.vala#L335

Do I proceed with the same solution?

**Comment by tintou** [comment]
Thanks for your patch, but we should never spawn a command line directly inside an application but use the dedicated API. There seem to be a D-Bus API for this that is on the GNOME Settings Daemon (that we are using). The equivalent command would be to enable BluetoothAirplaneMode as done here: https://gitlab.gnome.org/GNOME/gnome-settings-daemon/-/blob/gnome-3-28/plugins/rfkill/gsd-rfkill-manager.c#L33

**Review by tintou — DISMISSED** [review]
as in my comment above

**Comment by rafaelcorsi** [comment]
@tintou PR update, now using D-Bus. 

Inspired by: https://github.com/solus-project/budgie-desktop/blob/846c9334aa22394f5f3e24c0e6f6aa062122ebc5/src/applets/status/BluetoothIndicator.vala#L15

File: src/Services/Manager.vala [hunk 1]
```diff
@@ -215,6 +229,14 @@ public class BluetoothIndicator.Services.ObjectManager : Object {
                     }
                 }
             }
+            if (killer != null) {
+                killer.BluetoothAirplaneMode = true;
+            }
+        }
+        else {
```
Comment (danrabbit) [inline src/Services/Manager.vala:236]: elementary code style is cuddled else :) 
```suggestion
        } else {
```

File: src/Services/Manager.vala [hunk 2]
```diff
@@ -59,6 +66,13 @@ public class BluetoothIndicator.Services.ObjectManager : Object {
         }
 
         retrieve_finished = true;
+
+        this.setup_dbus_rfkill.begin (()=> {
+            if (this.killer == null) {
+              return;
+            }
+        });
```
Comment (danrabbit) [inline src/Services/Manager.vala:74]: 4 space indent, no need to use `this`
```suggestion
        setup_dbus_rfkill.begin (() => {
            if (killer == null) {
                return;
            }
        });
```

File: src/Services/Manager.vala [hunk 3]
```diff
@@ -28,6 +33,8 @@ public class BluetoothIndicator.Services.ObjectManager : Object {
     public bool is_powered {get; private set; default = false; }
     public bool is_connected {get; private set; default = false; }
 
+    Rfkill? killer = null;
```
Comment (danrabbit) [inline src/Services/Manager.vala:36]: Explicit scope please :)
```suggestion
    private Rfkill? killer = null;
```

File: src/Services/Manager.vala [hunk 4]
```diff
@@ -15,6 +15,11 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+[DBus (name="org.gnome.SettingsDaemon.Rfkill")]
+public interface Rfkill : Object {
```
Comment (danrabbit) [inline src/Services/Manager.vala:19]: This should probably be namespaced to make sure we don't conflict with other indicators, for example if the wifi indicator uses RFKill
```suggestion
public interface BluetoothIndicator.Services.Rfkill : Object {
```

File: src/Services/Manager.vala [hunk 5]
```diff
@@ -233,4 +255,16 @@ public class BluetoothIndicator.Services.ObjectManager : Object {
     public static bool compare_devices (Device device, Device other) {
         return device.modalias == other.modalias;
     }
+
+    private async void setup_dbus_rfkill () {
+        try {
+            killer = yield Bus.get_proxy (BusType.SESSION,
+                                          "org.gnome.SettingsDaemon.Rfkill",
+                                          "/org/gnome/SettingsDaemon/Rfkill");
```
Comment (danrabbit) [inline src/Services/Manager.vala:263]: elementary code style here is to treat parenthesis as braces

```suggestion
            killer = yield Bus.get_proxy (
                BusType.SESSION,
                "org.gnome.SettingsDaemon.Rfkill",
                "/org/gnome/SettingsDaemon/Rfkill"
            );
```

**Review by danrabbit — DISMISSED** [review]
Much cleaner over dbus :)

I left a few code style comments inline

File: src/Services/Manager.vala [hunk 6]
```diff
@@ -59,6 +66,13 @@ public class BluetoothIndicator.Services.ObjectManager : Object {
         }
 
         retrieve_finished = true;
+
+        setup_dbus_rfkill.begin (() => {
+            if (killer == null) {
+                return;
+            }
```
Comment (jeremypw) [inline src/Services/Manager.vala:73]: I do not see what this test does - there is no other code in the closure for when `killer != null` anyway??

File: src/Services/Manager.vala [hunk 7]
```diff
@@ -233,4 +254,18 @@ public class BluetoothIndicator.Services.ObjectManager : Object {
...
+    private async void setup_dbus_rfkill () {
+        try {
+            killer = yield Bus.get_proxy (
+                BusType.SESSION,
+                "org.gnome.SettingsDaemon.Rfkill",
+                "/org/gnome/SettingsDaemon/Rfkill"
+            );
+        } catch (Error e) {
+            killer = null;
+            return;
```
Comment (jeremypw) [inline src/Services/Manager.vala:267]: This `return` seems unnecessary - the function ends after this point anyway.   We would normally put a `debug` or `warning` message here if an error is unexpected.

**Comment by jeremypw** [comment]
Is this superceded by #129?

**Comment by leonardo-lemos** [comment]
> Is this superceded by #129?

I don't think so. #129 implements a Bus to rfkill and this PR uses the API from GSD as adviced.

**Comment by rafaelcorsi** [comment]
Dear all, @leonardo-lemos   contacted me by email to give continuity to this PR, it turns out that I had to migrate to ubuntu 20.04 and I am without eos and I have no way to validate it. 

Please feel free to modify the code as you wish, I hope to see this code on eos6.

all the best,

R. Corsi

File: src/Services/Manager.vala (hunk 7 above)
Comment (jeremypw) [inline src/Services/Manager.vala:267]:
```suggestion
```

File: src/Services/Manager.vala [hunk 8]
```diff
@@ -233,4 +254,17 @@ public class BluetoothIndicator.Services.ObjectManager : Object {
...
+            killer = yield Bus.get_proxy (
+                BusType.SESSION,
+                "org.gnome.SettingsDaemon.Rfkill",
+                "/org/gnome/SettingsDaemon/Rfkill"
+            );
+        } catch (Error e) {
+            killer = null;
+        }
+    }
+
```
Comment (jeremypw) [inline src/Services/Manager.vala:269]:
```suggestion
```

**Comment by jeremypw** [comment]
@tintou I cleaned up the code a little. Could you confirm whether this is OK to merge now?  Thanks.

**Comment by jeremypw** [comment]
Closing as old and no reviews for 2 years

**Pull request closed without merging**